### ISSUE TITLE
Find solution to allow starting to trace from any point in time

### ISSUE THREAD

**Issue by christophebedard** [ros2/ros2_tracing#44]
As mentioned in #40 (and [in the README](https://github.com/ros2/ros2_tracing#tracing)), we need to start tracing before the application launches to collect initialization information/metadata, otherwise the tracing data from the "runtime phase" (e.g., message publication data) is meaningless.

The current workaround is to start tracing before the application launches to collect the initialization data, then stop after 30-60 s, and then later on (when needed) start tracing again. This still requires us to actually set up tracing before the application launches. We would ideally like to be able to decide to enable tracing while an application is already running.

Some ideas:

1. Store initialization metadata and re-emit it at regular intervals. Then we'd simply need to record one metadata "burst." This does not sound very efficient, though.
    - This probably requires a bit of actual logic/code in `tracetools`.
2. Store initialization metadata (e.g., in simple caches) and "dump" it when we start tracing, which is similar to what LTTng does for its own stuff.
    - Look into LTTng "statedumps," which provide a dump of the current/initial state of the system so that a trace viewer/processor can make sense of the normal trace data that follows. 
    - This probably requires a bit of actual logic/code in `tracetools`.
3. Always trace the start of an application until some point (e.g., a function called by the user to declare that initialization is now over), possibly leveraging LTTng's [snapshot mode](https://lttng.org/docs/v2.13/#doc-taking-a-snapshot) to avoid writing the trace data to disk if it doesn't end up being needed (if we do not end up enabling tracing later).

This all would make live tracing more useful, so this is related to #149.

**Comment by ShravanDeva5327** [comment]
I am working on this as a part of [Improvements to ROS 2 Tracing](https://summerofcode.withgoogle.com/programs/2025/projects/quKpaaVA)

> Note: This is the initial workflow that @christophebedard and I have planned, we'll refine and update it as the project progresses.

**0. Defining the use case and requirements**

Situation we are trying to address:

- The user launches an application built with tracing instrumentation (which is enabled by default in ROS 2 now), but does not start tracing (neither with `ros2 trace` nor LTTng commands) before the application starts.
- Some time passes, that is, the initialization phase of ROS 2 completes.
- Eventually, the user wants to start tracing. This could be either:
a. To trace upcoming events, or
b. Because they notice something went wrong, in which case, they would also need tracing data from 3-4 minutes before the tracing was started (LTTng's flight recorder mode might help in this case).

Get input from other ROS 2 users and understand potential additional use cases of this feature.
- [x] Post on ROS Discourse to initiate the discussion

To support this use case, we need a way to obtain the initialization trace data so it can be used along with the current trace data (i.e., the data collected after the user started tracing) during analysis. Depending on the approach this might also involve merging the initialization and current trace data (it is important to note that there might be a large difference between the timestamps of initialization trace data and current trace data).

**1. Getting familiar with LTTng features**

- Try out different LTTng features with simple programs first, and eventually with ROS 2. Along the way, note down which features can potentially address our problem.

**2. Identifying the potential solutions**

- Ideate some potential solutions using the notes prepared from Step 1, and also explore the ideas mentioned in issue description.
- Email the [LTTng mailing list](https://lists.lttng.org/cgi-bin/mailman/listinfo/lttng-dev) explaining our use case and requirements, and ask for suggestions.
- Trying to find other solutions that might inspire or guide a solution for this specific use case

**3. Creating prototypes**

- Identify most promising and feasible solutions and create prototypes for each.

**4. Evaluating the prototypes**

- Evaluate the prototypes based on factors like:
a. Implementation
b. User experience
c. Performance/overhead

**5. Implementing the final solution**
- Select the best prototype, refine the implementation, add appropriate tests and documentation.

**Comment by christophebedard** [comment]
I had a thought yesterday regarding things to consider when evaluating potential solutions:

Given usual real-time contexts, where resources are created/allocated at the start of the application (i.e., initialization phase), I've always mostly assumed that nodes/publishers/subscriptions/etc. are all created during the start of the application. See Sections III-B and IV-A in the [`ros2_tracing` paper](https://arxiv.org/pdf/2201.00393). This is usually true, but there are definitely cases where publishers/subscriptions are destroyed and new ones are created as time passes. For example, tools like RViz and the CLI tools (e.g., `ros2 topic pub`), or any node that dynamically subscribes to topics. It's a bit annoying if the trace data is incomplete in cases like that. Since we've had to start tracing from the beginning anyway, we've always correctly collected initialization trace data about any pub/sub created later, so this was never really a problem. However, it becomes more important now.

In summary, when making it possible to start tracing at any point in time, we should keep in mind that the objects (nodes/publishers/subscriptions/etc.) that are active when we want to start tracing might not necessarily have been active when the application was started. This means that collecting "initialization" information about the currently-active objects _at the time where we start tracing_ would be more optimal than collecting initialization information _at the start of the application_ (in case it is needed later). That is why LTTng has its [state dump mechanism](https://lttng.org/docs/v2.13/#doc-regenerate-statedump).

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/seeking-feedback-on-enabling-post-launch-tracing-in-ros2-tracing/44130/1

**Comment by ShravanDeva5327** [comment]
> This is usually true, but there are definitely cases where publishers/subscriptions are destroyed and new ones are created as time passes. For example, tools like RViz and the CLI tools (e.g., ros2 topic pub), or any node that dynamically subscribes to topics.

I think we should consider labelling the tracepoints based on their types (initialization or runtime as in Table II of [ros2_tracing paper](https://arxiv.org/pdf/2201.00393)) and capture all the initialization events (whenever an initialization tracepoint is hit) irrespective of it happening within the first few minutes after the application starts or at any point before we start tracing. This way we don't have to bother about the objects that are active when we start tracing.

**Comment by christophebedard** [comment]
That is indeed one potential solution: _always_ trace/record initialization trace data. It would be a waste (overhead, storage) if it doesn't end up being used, but it probably wouldn't be a lot of data (unlike runtime trace data such as message publication). This specific behaviour could be opt-in. Also, the initialization trace data could potentially be kept in memory and only written to disk if it is needed (please check how to configure LTTng for this!).

Can you start maintaining a list of potential solutions and add this one?

**Comment by ShravanDeva5327** [comment]
We have explored 2 approaches to solve this problem:

1. **Dual Sessions**
Use two tracing sessions, one in snapshot mode for capturing initialization events, and another normal session for capturing all ROS 2 events. The normal tracing session would start when the user initiates tracing, while the snapshot session would be triggered automatically when a launch file is run. Users can opt in or out (from starting the snapshot session) via the launch file or an environment variable.

2. **Recording Initialization Events Alongside Runtime Events**
Instrument `rclcpp`, `rcl`, and `rmw` with a new set of tracepoints that collect the same initialization info but from within the runtime code, alongside existing runtime tracepoints. These tracepoints should only collect data once after tracing starts, ideally controlled via a boolean flag, to avoid duplicate events while preserving relative event order of initialization events.

**Comment by ShravanDeva5327** [comment]
We’ve decided to proceed with implementing the **Dual Sessions** prototype first (See #190)

**Comment by christophebedard** [comment]
Just a follow-up to:

> This means that collecting "initialization" information about the currently-active objects _at the time where we start tracing_ would be more optimal than collecting initialization information _at the start of the application_ (in case it is needed later). That is why LTTng has its [state dump mechanism](https://lttng.org/docs/v2.13/#doc-regenerate-statedump).

@ShravanDeva5327 you looked into possibly doing a ROS 2 state dump like LTTng does. One of the problems was that it was hard to trigger the initialization tracepoints on demand at runtime (i.e., during the state dump). Another problem was that there was no real way to trigger the state dump like LTTng does for itself internally (i.e., when it starts recording).

To get around this, you looked into triggering initialization tracepoints at runtime alongside the runtime tracepoints, which was still a bit tricky and not very straightforward, not to mention that the initialization tracepoints got triggered multiple times.

I attended the [Tracing Summit 2025](https://tracingsummit.org/) last week. There was a [presentation about LTTng (slides are available on the talk page)](https://cfp.tracingsummit.org/ts2025/talk/QZPAVL/). One of the upcoming features that was mentioned was "application state dump callbacks," which is what we were trying to do/use here.

That still wouldn't enable this solution, because, again, it's hard to trigger initialization tracepoints at runtime. Also, this is more of a long-term feature part of [libside](https://github.com/efficios/libside), which isn't yet used by LTTng-UST. However, I thought I'd mention it here.

**Comment by ros-discourse** [comment]
This issue has been mentioned on **Open Robotics Discourse**. There might be relevant details there:

https://discourse.openrobotics.org/t/gsoc-2025-improvements-to-ros-2-tracing/50062/1

**Comment by ros-discourse** [comment]
This issue has been mentioned on **Open Robotics Discourse**. There might be relevant details there:

https://discourse.openrobotics.org/t/gsoc-2025-improvements-to-ros-2-tracing/50063/1
